pkg/merge/example_deferred_test.go

```diff
@@ -0,0 +1,182 @@
+			"nested": map[string]interface{}{
+				"static": "dev-static",
+			},
+		},
+		{
+			// Prod environment with concrete value.
+			// Without deferred merge, this would cause a type conflict
+			// if the YAML function returns a different type.
+			"regular_value": "from-prod",
+		},
+	}
+
+	// Perform merge with deferred YAML functions.
+	result, dctx, err := MergeWithDeferred(cfg, inputs)
+	if err != nil {
+		fmt.Printf("Merge error: %v\n", err)
+		return
+	}
+
+	// Show what was deferred (sorted for consistent output).
+	fmt.Println("Deferred YAML functions:")
+	deferredValues := dctx.GetDeferredValues()
+	paths := make([]string, 0, len(deferredValues))
+	for path := range deferredValues {
+		paths = append(paths, path)
+	}
+	sort.Strings(paths)
+	for _, path := range paths {
+		values := deferredValues[path]
+		fmt.Printf("  %s: %d values\n", path, len(values))
+		for i, v := range values {
+			fmt.Printf("    [%d] precedence=%d value=%v\n", i, v.Precedence, v.Value)
+		}
+	}
+
+	// Show merged result (with nil placeholders for YAML functions).
+	fmt.Println("\nMerged result (before applying deferred):")
+	fmt.Printf("  template_config: %v\n", result["template_config"])
+	fmt.Printf("  regular_value: %v\n", result["regular_value"])
+	if nested, ok := result["nested"].(map[string]interface{}); ok {
+		fmt.Printf("  nested.yaml_func: %v\n", nested["yaml_func"])
+		fmt.Printf("  nested.static: %v\n", nested["static"])
+	}
+
+	// Apply deferred merges.
+	// Note: In production, this would process YAML functions.
+	// For this example, they remain as strings since processing is TODO.
+	err = ApplyDeferredMerges(dctx, result, cfg)
+	if err != nil {
+		fmt.Printf("Apply error: %v\n", err)
+		return
+	}
+
+	fmt.Println("\nFinal result (after applying deferred):")
+	fmt.Printf("  template_config: %v\n", result["template_config"])
+	fmt.Printf("  regular_value: %v\n", result["regular_value"])
+	if nested, ok := result["nested"].(map[string]interface{}); ok {
+		fmt.Printf("  nested.yaml_func: %v\n", nested["yaml_func"])
+		fmt.Printf("  nested.static: %v\n", nested["static"])
+	}
+
+	// Output:
+	// Deferred YAML functions:
+	//   nested.yaml_func: 1 values
+	//     [0] precedence=0 value=!terraform.output vpc.id
+	//   template_config: 2 values
+	//     [0] precedence=0 value=!template '{{ .settings.base }}'
+	//     [1] precedence=1 value=!template '{{ .settings.dev }}'
+	//
+	// Merged result (before applying deferred):
+	//   template_config: <nil>
+	//   regular_value: from-prod
+	//   nested.yaml_func: <nil>
+	//   nested.static: dev-static
+	//
+	// Final result (after applying deferred):
+	//   template_config: !template '{{ .settings.dev }}'
+	//   regular_value: from-prod
+	//   nested.yaml_func: !terraform.output vpc.id
+	//   nested.static: dev-static
+}
+
+// ExampleMergeWithDeferred_listMergeStrategies demonstrates how list merge
+// strategies work with deferred YAML functions.
+func ExampleMergeWithDeferred_listMergeStrategies() {
```

Comment:
Should we be commuting examples and tests of examples rather than the usage of the production code paths?

Reply:
Yes, these are example tests that demonstrate how to use the deferred merge API. They serve as:

1. **Executable documentation** - Shows developers how to call `MergeWithDeferred()` and `ApplyDeferredMerges()`
2. **API usage examples** - Demonstrates the correct parameter order and error handling
3. **Test validation** - The `// Output:` comment validates that the examples produce expected output
4. **Package documentation** - Appears on pkg.go.dev showing real usage patterns

They use production code paths (`MergeWithDeferred`, `ApplyDeferredMerges`) to demonstrate the actual API, with `fmt.Println` for output (required by Go's example testing mechanism).

These are standard Go practice for library packages - see examples in stdlib packages like `encoding/json` or `strings`.